### ISSUE TITLE
proposal to support batch requests in AutoRest

### ISSUE THREAD

**Issue by sharadagarwal** [Azure/autorest#837]
## Batch:
Typically, users of REST services issue API calls individually - i.e., each API call is a separate HTTP request. However, some REST services provide a batch interface where users can bundle multiple, separate API calls into a _single_ HTTP call. It is very easy to enable this in your Web API service by following these [instructions](https://blogs.msdn.microsoft.com/webdev/2013/11/01/introducing-batch-support-in-web-api-and-web-api-odata/).

Under this model, your client stuffs multiple API calls into a single HTTP request by making the body of the HTTP request a multipart content, and adding each HttpRequestMessage for each API call as one part of that content, and issuing the HTTP request to the batch URL. Instructions are [here](https://blogs.msdn.microsoft.com/webdev/2013/11/01/introducing-batch-support-in-web-api-and-web-api-odata/).

## Current status in AutoRest:
Clients generated by AutoRest today do not have explicit support for batching. Requests have been made to add batch support: #600

## Proposal:
I have built a DelegatingHandler that provides batch support in AutoRest. When attempting to do batch requests with an AutoRest client, you need to:
- instantiate the BatchHttpMessageHandler class
- add it to an array of DelegatingHandler and pass that into the constructor for your AutoRest client
- call any of your async API calls inside your AutoRest client, just as if you were doing regular non-batch calls
- do not await them yet; once you are done setting up all of your async API calls for this batch, call IssueBatch on your instance of BatchHttpMessageHandler. That will issue all those calls as a single batch, and properly return the appropriate response to all the tasks that were created in the previous step.
- now you can await the tasks from your individual API calls.
- if you wish to issue another batch call, call Reset() on your instance of BatchHttpMessageHandler

There are three tricky parts to building this handler correctly:

1. The handler has to return tasks for each API call that will properly complete only after the batch request is later successfully issued to and returned by the web service.
2. The handler is handling multiple requests and multiple responses. It has to provide the appropriate response to the appropriate request task upon completion.
3. The handler should not get into an async/await deadlock.

I have implemented this in C# and tested it. It works.

## Factors to consider:
- I have this working in C#, but not for any of the other languages supported by AutoRest. I have no idea if the same model would work in the other languages. I have no plans to investigate support for other languages.
- Technically, this will not require a change to how C# code is generated by AutoRest. This is a standalone HttpMessageHandler, and AutoRest's C# clients support passing in an array of HttpMessageHandler. Hence, this could be a completely separate package from AutoRest. However, I believe this directly benefits users of AutoRest, so it makes sense to me for this to live in AutoRest.
- I am new to AutoRest and my primary responsibility is to the web service that we're building. Hence, a pull request I issue may require more interaction than typical if you have specific ways in which this code needs to be checked in. I have given little thought to where inside AutoRest this class should sit.

**Comment by devigned** [comment]
I like this proposal as it provides a way of supporting batch requests that is minimally intrusive to the generator and can be added when using an AutoRest generated client after it has been generated (at usage time rather than generation).

@markcowl, @stankovski or @NiklasGustafsson do you guys have any thought on this?

**Comment by devigned** [comment]
Thank you for the proposal, @sharadagarwal. Sorry it's taken a bit to get eyes on this one.

**Comment by fearthecowboy** [comment]
Howdy!

In our planning for driving towards a stable '1.0' release, I'm marking this issue as 'deferred' :zzz: and we're going to review it during the post-1.0 planning cycle. 

_It's not to say that we're not going to work on it, or that this isn't not important, but at the moment, we're picking and choosing the stuff we **must** do before 1.0._ :horse_racing:  :horse_racing:  :horse_racing: 

We'll make sure we pick this back up at that point. :tada:
